### DISCUSSION TITLE
How to Fix my Character Controller

### DISCUSSION THREAD

**Discussion by cuppajoeman** [jrouwe/JoltPhysics#984] [Q&A] [answered]
Hi, I recently built [a project](https://github.com/opengl-toolbox/mwe_physics_world_with_character) in isolation from my game, I want to use this project to make the character controller feel right. 

As of right now I am using a `CharacterVirtual` without much configuration. I also found out that godot has a plugin to switch to jolt, so I compared the two: 

https://streamable.com/axfb3y

As you can see there are various unwanted things that occur with the pure `CharacterVirtual` implementation, such as:
- "locking in" to edges
- "bouncing" when sitting in a V shape wedge
- gravity applying weirdly when falling off an object (not jumping off) 
- just generally doesn't feel smooth
- being able to "wall jump"  (I think it's detecting on-ground state when it shouldn't)
- the jump height doesn't feel consistent (I do a simple on-ground state check, maybe it's activated for longer than I think?), I'd have to actually test this first.

---

Do I have any issues with my code which are causing these problems? Here are the relevant files:
- [create_character()](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/8b2497f13a0d72d3dbff3d4736abbb89d370cf3d/interaction/physics/physics.cpp#L130)
- [update](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/8b2497f13a0d72d3dbff3d4736abbb89d370cf3d/main.cpp#L139) - contains jump logic and gravity application.

---

The version of this project where the jolt-godot plugin is used feels extremely good out of the box and doesn't have many glaring issues in comparsion.

I looked at the [godot-jolt](https://github.com/godot-jolt/godot-jolt) and couldn't find any relevant info about a character controller. From what I understand, godot will use their own character controller logic [defined here](https://github.com/godotengine/godot/blob/master/scene/3d/physics/character_body_3d.cpp) which just sits atop jolt. 

If these issues are not related to my code, then maybe there's things to learn from the godot implementation? I'd be willing to help contribute if improvements to the character controller (virtual) are in scope.

**Comment by jrouwe** [comment]
1. You're using a cylinder shape for the character, please switch to a capsule shape instead (much more stable).
2. You're not using CharacterBaseSettings::mSupportingVolume which means any contact point can be considered a ground contact (hence you can wall jump)

For an example please look at the CharacterVirtual demo:

https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterVirtualTest.cpp#L37-L160

If you have specific issues, please modify this test to show the problem.

**Reply by cuppajoeman** [reply]
Hey, I would love to use the included samples to test out functionality, but I use linux, and my hardware doesn't support vulkan, which implies that I cannot run any of the samples (the wine solution requires that my hardware supports vulkan). 

That's partially why I started this project; so that I could get basic things working in Jolt like this character.

---

 I made the relevant changes in my code (changed to cylinder shape and added a supporting volume). Things are still a bit wonky: 

https://streamable.com/yi4tvj

I cross checked how the demo does jumping and how I do jumping and the logic seems quite similar: [my jumping](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/main.cpp#L182-L188),  [your jumping](https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterVirtualTest.cpp#L123-L135).

This is [how I apply gravity](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/c25f6898390d7b48daef7b87b87c6722442a21c8/main.cpp#L178), and [this is how the sample does it](https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterVirtualTest.cpp#L140).

The only other major differences is that I can see are that the sample seems to [implement air-strafing capabilities](https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterVirtualTest.cpp#L95-L108), it supports a crouching state, and implements `PrePhysicsUpdate`, which [optionally enables floor sticking and stair stepping](https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterVirtualTest.cpp#L60-L69). My project does not use air-strafing, crouching or floor-sticking/stair-stepping.

---

My project is really a minimal working example, it sets up Jolt by following the structure used in the[ `HelloWorld.cpp` example](https://github.com/jrouwe/JoltPhysics/blob/master/HelloWorld/HelloWorld.cpp) in the file [`jolt_implementation.hpp`](https://github.com/opengl-toolbox/mwe_physics_world_with_character/blob/main/interaction/physics/jolt_implementation.hpp) then creates `VirtualCharacter` and loads in the world mesh by creating a `MeshShape` for each component of the map and putting it in the non-moving layer.

It creates a simple GLFW window, uses OpenGL3.3 for graphics, implements the debug renderer and simply uses keypresses to drive the virtual character and rotate it with mouse input. The only relevant files for physics would be `main.cpp` (gets keypresses and drives the character) and `physics.cpp` (sets up the character, runs physics updates). 

The only other variable I can think of would be the [various settings](https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterVirtualTest.cpp#L24-L32) used for the character in the sample, in my code I only set up a `Shape` and a `SupportingVolume`.

Also I have no reference to how character virtual in the demo behaves, I assume it's similar to the [regular character demo](https://www.youtube.com/watch?v=YjaJT9of7UE) ?

**Reply by jrouwe** [reply]
It's quite important to have stair stepping turned on. Another thing that I can think of is that you're not using meters as units, the default settings have been tweaked for a character that's roughly 2 m tall.

Other than that I don't know what is wrong with your project. I really suggest you find someone with a Windows PC and create a repro in the Samples app.

**Reply by cuppajoeman** [reply]
I just went out and bought windows so I can test this myself, I'll see if I can reproduce the behavior over the next week.

**Reply by cuppajoeman** [reply]
@jrouwe Now that I can at least run the samples, and have been able to move around the sample map. Just to confirm, if I want to add obstacles to the map such as spiky pyramids to step on [I can edit here ?](https://github.com/jrouwe/JoltPhysics/blob/de3517e9ce5790eec2111370d74bd3b21873437a/Samples/Tests/Character/CharacterBaseTest.cpp#L111)

**Reply by jrouwe** [reply]
Yes

**Reply by cuppajoeman** [reply]
Hi @jrouwe I got around to making a minimal working example for the strange behavior with the virtual character controller. 

In this implemention I simply use a constant acceleration with friction to move the player around I have no settings set besides the shape and the supporting body. 

I realize that I could have simply copied many of the the things you use in your default implementation, but my game's movement doesn't work exactly like the one specified in the example. 

Also I didn't enable very many of the settings/custom logic for the character, because I'd like to be able to understand when and why we need to enable certain settings and what problems they fix. With that knowledge I would like to improve the docs for using the virtual character controller.

To test this MWE use this as source for `CharacterVirtualTest.cpp`
```cpp
// Jolt Physics Library (https://github.com/jrouwe/JoltPhysics)
// SPDX-FileCopyrightText: 2021 Jorrit Rouwe
// SPDX-License-Identifier: MIT

#include <TestFramework.h>

#include <Tests/Character/CharacterVirtualTest.h>
#include <Jolt/Physics/Collision/Shape/CapsuleShape.h>
#include <Jolt/Physics/Collision/Shape/RotatedTranslatedShape.h>
#include <Layers.h>
#include <Renderer/DebugRendererImp.h>
#include <Application/DebugUI.h>

JPH_IMPLEMENT_RTTI_VIRTUAL(CharacterVirtualTest)
{
	JPH_ADD_BASE_CLASS(CharacterVirtualTest, CharacterBaseTest)
}

void CharacterVirtualTest::Initialize()
{
	CharacterBaseTest::Initialize();

	// Create 'player' character
	Ref<CharacterVirtualSettings> settings = new CharacterVirtualSettings();
	settings->mShape = mStandingShape;
	settings->mSupportingVolume = Plane(Vec3::sAxisY(), -cCharacterRadiusStanding); // Accept contacts that touch the lower sphere of the capsule
	mCharacter = new CharacterVirtual(settings, RVec3::sZero(), Quat::sIdentity(), 0, mPhysicsSystem);
	mCharacter->SetListener(this);
}

void CharacterVirtualTest::PrePhysicsUpdate(const PreUpdateParams &inParams)
{
	CharacterBaseTest::PrePhysicsUpdate(inParams);

	// Draw character pre update (the sim is also drawn pre update)
	RMat44 com = mCharacter->GetCenterOfMassTransform();
	RMat44 world_transform = mCharacter->GetWorldTransform();
#ifdef JPH_DEBUG_RENDERER
	mCharacter->GetShape()->Draw(mDebugRenderer, com, Vec3::sReplicate(1.0f), Color::sGreen, false, true);
#endif // JPH_DEBUG_RENDERER

	// Draw shape including padding (only implemented for capsules right now)
	if (static_cast<const RotatedTranslatedShape *>(mCharacter->GetShape())->GetInnerShape()->GetSubType() == EShapeSubType::Capsule)
	{
		if (mCharacter->GetShape() == mStandingShape)
			mDebugRenderer->DrawCapsule(com, 0.5f * cCharacterHeightStanding, cCharacterRadiusStanding + mCharacter->GetCharacterPadding(), Color::sGrey, DebugRenderer::ECastShadow::Off, DebugRenderer::EDrawMode::Wireframe);
		else
			mDebugRenderer->DrawCapsule(com, 0.5f * cCharacterHeightCrouching, cCharacterRadiusCrouching + mCharacter->GetCharacterPadding(), Color::sGrey, DebugRenderer::ECastShadow::Off, DebugRenderer::EDrawMode::Wireframe);
	}

	// Remember old position
	RVec3 old_position = mCharacter->GetPosition();

	// Settings for our update function
	CharacterVirtual::ExtendedUpdateSettings update_settings;

	// Update the character position
	mCharacter->ExtendedUpdate(inParams.mDeltaTime,
		-mCharacter->GetUp() * mPhysicsSystem->GetGravity().Length(),
		update_settings,
		mPhysicsSystem->GetDefaultBroadPhaseLayerFilter(Layers::MOVING),
		mPhysicsSystem->GetDefaultLayerFilter(Layers::MOVING),
		{ },
		{ },
		*mTempAllocator);

	// Calculate effective velocity
	RVec3 new_position = mCharacter->GetPosition();
	Vec3 velocity = Vec3(new_position - old_position) / inParams.mDeltaTime;

	// Draw state of character
	DrawCharacterState(mCharacter, world_transform, velocity);

	// Draw labels on ramp blocks
	for (size_t i = 0; i < mRampBlocks.size(); ++i)
		mDebugRenderer->DrawText3D(mBodyInterface->GetPosition(mRampBlocks[i]), StringFormat("PushesPlayer: %s\nPushable: %s", (i & 1) != 0? "True" : "False", (i & 2) != 0? "True" : "False"), Color::sWhite, 0.25f);
}

void CharacterVirtualTest::HandleInput(Vec3Arg inMovementDirection, bool inJump, bool inSwitchStance, float inDeltaTime)
{
    const int movement_acceleration = 15;
    const int jump_impulse_force = 1000;
    const float friction = 0.983f;

    Vec3 updated_velocity = mCharacter->GetLinearVelocity();
    updated_velocity += movement_acceleration * inMovementDirection * inDeltaTime;

    updated_velocity *= friction;

    updated_velocity += mPhysicsSystem->GetGravity() * inDeltaTime;

    if (inJump && mCharacter->GetGroundState() == CharacterVirtual::EGroundState::OnGround) {
      updated_velocity += jump_impulse_force * mCharacter->GetUp() * inDeltaTime;
    }

    mCharacter->SetLinearVelocity(updated_velocity);

}

void CharacterVirtualTest::AddCharacterMovementSettings(DebugUI* inUI, UIElement* inSubMenu)
{
	inUI->CreateCheckBox(inSubMenu, "Enable Character Inertia", sEnableCharacterInertia, [](UICheckBox::EState inState) { sEnableCharacterInertia = inState == UICheckBox::STATE_CHECKED; });
}

void CharacterVirtualTest::AddConfigurationSettings(DebugUI *inUI, UIElement *inSubMenu)
{
	inUI->CreateComboBox(inSubMenu, "Back Face Mode", { "Ignore", "Collide" }, (int)sBackFaceMode, [=](int inItem) { sBackFaceMode = (EBackFaceMode)inItem; });
	inUI->CreateSlider(inSubMenu, "Up Rotation X (degrees)", RadiansToDegrees(sUpRotationX), -90.0f, 90.0f, 1.0f, [](float inValue) { sUpRotationX = DegreesToRadians(inValue); });
	inUI->CreateSlider(inSubMenu, "Up Rotation Z (degrees)", RadiansToDegrees(sUpRotationZ), -90.0f, 90.0f, 1.0f, [](float inValue) { sUpRotationZ = DegreesToRadians(inValue); });
	inUI->CreateSlider(inSubMenu, "Max Slope Angle (degrees)", RadiansToDegrees(sMaxSlopeAngle), 0.0f, 90.0f, 1.0f, [](float inValue) { sMaxSlopeAngle = DegreesToRadians(inValue); });
	inUI->CreateSlider(inSubMenu, "Max Strength (N)", sMaxStrength, 0.0f, 500.0f, 1.0f, [](float inValue) { sMaxStrength = inValue; });
	inUI->CreateSlider(inSubMenu, "Character Padding", sCharacterPadding, 0.01f, 0.5f, 0.01f, [](float inValue) { sCharacterPadding = inValue; });
	inUI->CreateSlider(inSubMenu, "Penetration Recovery Speed", sPenetrationRecoverySpeed, 0.0f, 1.0f, 0.05f, [](float inValue) { sPenetrationRecoverySpeed = inValue; });
	inUI->CreateSlider(inSubMenu, "Predictive Contact Distance", sPredictiveContactDistance, 0.01f, 1.0f, 0.01f, [](float inValue) { sPredictiveContactDistance = inValue; });
	inUI->CreateCheckBox(inSubMenu, "Enable Walk Stairs", sEnableWalkStairs, [](UICheckBox::EState inState) { sEnableWalkStairs = inState == UICheckBox::STATE_CHECKED; });
	inUI->CreateCheckBox(inSubMenu, "Enable Stick To Floor", sEnableStickToFloor, [](UICheckBox::EState inState) { sEnableStickToFloor = inState == UICheckBox::STATE_CHECKED; });
}

void CharacterVirtualTest::SaveState(StateRecorder &inStream) const
{
	CharacterBaseTest::SaveState(inStream);

	mCharacter->SaveState(inStream);

	bool is_standing = mCharacter->GetShape() == mStandingShape;
	inStream.Write(is_standing);

	inStream.Write(mAllowSliding);
	inStream.Write(mDesiredVelocity);
}

void CharacterVirtualTest::RestoreState(StateRecorder &inStream)
{
	CharacterBaseTest::RestoreState(inStream);

	mCharacter->RestoreState(inStream);

	bool is_standing = mCharacter->GetShape() == mStandingShape; // Initialize variable for validation mode
	inStream.Read(is_standing);
	mCharacter->SetShape(is_standing? mStandingShape : mCrouchingShape, FLT_MAX, { }, { }, { }, { }, *mTempAllocator);

	inStream.Read(mAllowSliding);
	inStream.Read(mDesiredVelocity);
}

void CharacterVirtualTest::OnAdjustBodyVelocity(const CharacterVirtual *inCharacter, const Body &inBody2, Vec3 &ioLinearVelocity, Vec3 &ioAngularVelocity)
{
	// Apply artificial velocity to the character when standing on the conveyor belt
//	if (inBody2.GetID() == mConveyorBeltBody)
//		ioLinearVelocity += Vec3(0, 0, 2);
}

void CharacterVirtualTest::OnContactAdded(const CharacterVirtual *inCharacter, const BodyID &inBodyID2, const SubShapeID &inSubShapeID2, RVec3Arg inContactPosition, Vec3Arg inContactNormal, CharacterContactSettings &ioSettings)
{
	// Draw a box around the character when it enters the sensor
	if (inBodyID2 == mSensorBody)
	{
		AABox box = inCharacter->GetShape()->GetWorldSpaceBounds(inCharacter->GetCenterOfMassTransform(), Vec3::sReplicate(1.0f));
		mDebugRenderer->DrawBox(box, Color::sGreen, DebugRenderer::ECastShadow::Off, DebugRenderer::EDrawMode::Wireframe);
	}

//	// Dynamic boxes on the ramp go through all permutations
//	Array<BodyID>::const_iterator i = find(mRampBlocks.begin(), mRampBlocks.end(), inBodyID2);
//	if (i != mRampBlocks.end())
//	{
//		size_t index = i - mRampBlocks.begin();
//		ioSettings.mCanPushCharacter = (index & 1) != 0;
//		ioSettings.mCanReceiveImpulses = (index & 2) != 0;
//	}
//
//	// If we encounter an object that can push us, enable sliding
//	if (ioSettings.mCanPushCharacter && mPhysicsSystem->GetBodyInterface().GetMotionType(inBodyID2) != EMotionType::Static)
//		mAllowSliding = true;
}

void CharacterVirtualTest::OnContactSolve(const CharacterVirtual *inCharacter, const BodyID &inBodyID2, const SubShapeID &inSubShapeID2, RVec3Arg inContactPosition, Vec3Arg inContactNormal, Vec3Arg inContactVelocity, const PhysicsMaterial *inContactMaterial, Vec3Arg inCharacterVelocity, Vec3 &ioNewCharacterVelocity)
{
//	// Don't allow the player to slide down static not-too-steep surfaces when not actively moving and when not on a moving platform
//	if (!mAllowSliding && inContactVelocity.IsNearZero() && !inCharacter->IsSlopeTooSteep(inContactNormal))
//		ioNewCharacterVelocity = Vec3::sZero();
}
```

With this setup you'll see that the virtual character controller seems to "twitch" around a lot more, especially when it comes to the ramps, when you try to go onto the ramp the character tends to get stuck in a similar way that occurs in my project.

https://streamable.com/huzf9i

**Comment by jrouwe** [comment, marked as answer]
Hello,

The problem with your solution is that you never reset the vertical velocity of the character (unless you jump). If you add:

```
mDebugRenderer->DrawArrow(mCharacter->GetPosition(), mCharacter->GetPosition() + updated_velocity, Color::sRed, 0.01f);
```

at the end of `CharacterVirtualTest::HandleInput` you'll see that the `updated_velocity` vector is ever increasing. If we for example go to the ramp:

![image](https://github.com/jrouwe/JoltPhysics/assets/1621693/ced64f37-4097-48f6-a880-a20ad640c7d5)

You'll see that moving into this ramp results in a tiny bit of movement towards the ramp and a very large movement towards the floor. So you'll slide down the ramp rather than up.

This is why the example separates the velocity in a horizontal and a vertical velocity and resets the vertical velocity when you're moving into the floor.

**Reply by cuppajoeman** [reply]
Hey @jrouwe  that worked really well, I ended up changing the code to be: 

```cpp
void CharacterVirtualTest::HandleInput(Vec3Arg inMovementDirection, bool inJump, bool inSwitchStance, float inDeltaTime)
{
    const int movement_acceleration = 15;
    const int jump_impulse_force = 1000;
    const float friction = 0.983f;

    Vec3 updated_velocity = mCharacter->GetLinearVelocity();
    updated_velocity += movement_acceleration * inMovementDirection * inDeltaTime;

    updated_velocity *= friction;

    if (mCharacter->GetGroundState() == CharacterVirtual::EGroundState::OnGround) {
      updated_velocity.SetY(0.0f);
      if (inJump) {
          updated_velocity += jump_impulse_force * mCharacter->GetUp() * inDeltaTime;
      }
    }

    updated_velocity += mPhysicsSystem->GetGravity() * inDeltaTime;


    mCharacter->SetLinearVelocity(updated_velocity);

}
```

Would it be within scope to add this code as a minimal working implementation for the `VirtualCharacter` in the docs where we mention that you need to at least bind a `mShape` and `mSupporting` volume for the character to get it working as expected?

**Reply by jrouwe** [reply]
No, I think the examples provide a better minimal example. Your code here doesn't take moving platforms into account. If you were to jump on one you'd slide off as if it were ice. Also your code doesn't support surface velocity.

**Reply by cuppajoeman** [reply]
Ok, sounds good.
